dpath/util.py

```diff
@@ -172,48 +232,74 @@ def merge(dst, src, separator="/", afilter=None, flags=MERGE_ADDITIVE, _path="")
         * MERGE_TYPESAFE : When 2 keys at equal levels are of different
           types, raise a TypeError exception. By default, the source
           replaces the destination in this situation.
-    """
+    '''
+    filtered_src = search(src, '**', afilter=afilter, separator='/')
 
-    if afilter:
-        # Having merge do its own afiltering is dumb, let search do the
-        # heavy lifting for us.
-        src = search(src, '**', afilter=afilter)
-        return merge(dst, src)
+    def merger(dst, src, _segments=()):
+        for key, found in dpath.segments.kvs(src):
+            # Our current path in the source.
+            segments = _segments + (key,)
 
-    def _check_typesafe(obj1, obj2, key, path):
-        if not key in obj1:
-            return
-        elif ( (flags & MERGE_TYPESAFE == MERGE_TYPESAFE) and (type(obj1[key]) != type(obj2[key]))):
-            raise TypeError("Cannot merge objects of type {0} and {1} at {2}"
-                            "".format(type(obj1[key]), type(obj2[key]), path))
-        elif ( (flags & MERGE_TYPESAFE != MERGE_TYPESAFE) and (type(obj1[key]) != type(obj2[key]))):
-            obj1.pop(key)
-
-    if isinstance(src, MutableMapping):
-        for (i, v) in enumerate(src):
-            _check_typesafe(dst, src, v, separator.join([_path, str(v)]))
-
-            if not v in dst:
-                dst[v] = src[v]
-            else:
-                if not isinstance(src[v], (MutableMapping, MutableSequence)):
-                    dst[v] = src[v]
-                else:
-                    merge(dst[v], src[v], afilter=afilter, flags=flags,
-                          _path=separator.join([_path, str(v)]), separator=separator)
-    elif isinstance(src, MutableSequence):
-        for (i, v) in enumerate(src):
-            _check_typesafe(dst, src, i, separator.join([_path, str(i)]))
-            dsti = i
-            if ( flags & MERGE_ADDITIVE):
-                dsti = len(dst)
-            if dsti >= len(dst):
-                dst += [None] * (dsti - (len(dst) - 1))
-            if dst[dsti] == None:
-                dst[dsti] = src[i]
-            else:
-                if not isinstance(src[i], (MutableMapping, MutableSequence)):
-                    dst[dsti] = src[i]
-                else:
-                    merge(dst[i], src[i], afilter=afilter, flags=flags,
-                          _path=separator.join([_path, str(i)]), separator=separator)
+            if len(key) == 0 and not options.ALLOW_EMPTY_STRING_KEYS:
```

Comment:
Will fail on integer keys, need to use the same try: lenght = len(key) trick used in walk().